ec2 architecture and guide (#4585)
camunda/camunda-docs: docs/self-managed/setup/deploy/amazon/aws-ec2.md

  - a [Private Subnet](https://docs.aws.amazon.com/vpc/latest/userguide/configure-subnets.html), which does not have direct access to the internet and can not easily be reached.

Both types of subnets are distributed over three availability zones of a single AWS region, allowing for a highly available setup.

The private subnet contains:

- three [EC2](https://aws.amazon.com/ec2/) instances, which will run Camunda 8.
- a [managed OpenSearch](https://aws.amazon.com/what-is/opensearch/) cluster stretched over the three availability zones.

The public subnet contains:

- (optional) an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB) is used to expose the WebUIs like Operate, Tasklist, and Connectors to the outside world. This is done using sticky sessions as generally requests are distributed round-robin across all EC2 instances. This component is stretched over all three availability zones.

> - (optional) a [Network Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) (NLB) is used to expose the gRPC endpoint of the Zeebe Gateway. This will eventually be superseded and integrated via the unified REST API into the ALB. This component is stretched over all three availability zones.

Reviewer: good to know, it always sounded more like gRPC will remain for internal usages.

- (optional) a [Bastion Host](https://en.wikipedia.org/wiki/Bastion_host) to allow access to the private EC2 instances since they're not publicly exposed.
  - Alternatively, one can utilize the [AWS Client VPN](https://docs.aws.amazon.com/vpn/latest/clientvpn-admin/what-is.html) instead to reach the private subnet within the VPC. The setup requires a lot of extra work and certificates, but one can follow the [getting started tutorial by AWS](https://docs.aws.amazon.com/vpn/latest/clientvpn-admin/cvpn-getting-started.html).
- a NAT Gateway that allows the private EC2 instances to reach the internet to download and update software packages. This can not be used to access the EC2 instances.

Lastly, an [Internet Gateway](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Internet_Gateway.html) as glue component between the VPC and the Internet, to allow traffic between those.

## Prerequisites

> - An AWS account to create any resources within AWS.

Reviewer: I'll add it with a high-level in the docs and add a more minimal required policy to the repo as a reference to link to.
https://github.com/iann0036/iamlive helps to generate it but would be too detailed for the docs.

Apply the Terraform configuration to create the resources.

```sh
terraform apply
```

The execution takes roughly 30 minutes of which the majority of time ~ 25 minutes is solely the creation of a managed highly available OpenSearch cluster.

6. Access outputs

After the infrastructure is created, you can access the outputs defined in `outputs.tf`. For example, to get the OpenSearch endpoint:

```sh
terraform output aws_opensearch_domain
```

7. (optional) Connect to remote machines via Bastion host

The EC2 instances are not public and have to be reached via a Bastion host.
Alternatively one can utilize the [AWS VPN Client](https://docs.aws.amazon.com/vpn/latest/clientvpn-admin/what-is.html) to connect securely to a private VPC. This is not covered here as the setup requires a lot more manual user interaction.

```
export BASTION_HOST=$(terraform output -raw bastion_ip)
# retrieves the first IP from the camunda_ips array
export CAMUNDA_IP=$(tf output -json camunda_ips | jq -r '.[0]')

ssh -J admin@${BASTION_HOST} admin@${CAMUNDA_IP}
```

### Deploy Camunda 8

1. Go to directory

```
cd camunda-deployment-references/aws/ec2/scripts
```

2. Consider optional features

> There are certain features hidden behind a feature flag. Those are the following:

Reviewer: I'd call them script features, they have nothing to do with Terraform and very little with C8. (e.g security is C8 bound).

- `CLOUDWATCH_ENABLED`: The default is false. If set to true will install the CloudWatch agent on each EC2 instance and export Camunda logs and Prometheus metrics to AWS CloudWatch.

> - `SECURITY`: The default is false. If set to true will use self-signed certificates to secure cluster communication, based on the procedure described in the [documentation](/self-managed/zeebe-deployment/security/secure-cluster-communication.md). This requires a manual step as a prerequisite as described below in step 3.

Reviewer: No, Cloudwatch can be quite expensive and maybe companies already have their own monitoring stack.
For security no as well, self-sined certificates is meant for development purposes.

3. Execute script `all-in-one-install.sh`

Run the `all-in-one-install.sh` script.

It will install all required dependencies. Additionally, it configures Camunda 8 to run in a highly available setup by using a managed OpenSearch instance.

The script will pull all required IPs and other information from the Terraform state via Terraform outputs.

During the first installation, you will be asked to confirm the connection to each EC2 instance by typing `yes`.

4. Use the setup

Via the Terraform output `alb_endpoint` one can access Operate or on port `9090` the Connectors instance.

Via the Terraform output `nlb_endpoint` one can access the gRPC endpoint of the Zeebe gateway.

The two endpoints above use the publicly assigned hostname of AWS. One may add their domain via CNAME records or use [Route53](https://aws.amazon.com/route53/) to map to the load balancers, allowing them to easily enable SSL. This will require extra work in the Terraform blueprint as it listens to HTTP by default.

> Alternatively, if you have decided not to expose your environment, you can use the jump host to access relevant services on your local machine via port-forwarding.

Reviewer: I mentioned it under the Bastion host as alternative but I can just reiterate on it as alternative for enterprise grade setups.

The following can be executed from within the Terraform folder to bind the remote ports to your local machine.

```
export BASTION_HOST=$(terraform output -raw bastion_ip)
# retrieves the first IP from the camunda_ips array
export CAMUNDA_IP=$(tf output -json camunda_ips | jq -r '.[0]')

# 26500 - gRPC; 8080 - WebUI; 9090 - Connectors
ssh -L 26500:${CAMUNDA_IP}:26500 -L 8080:${CAMUNDA_IP}:8080 -L 9090:${CAMUNDA_IP}:9090 admin@${BASTION_HOST}
```

5: (optional) turn off bastion host

> ### Upgrade Camunda 8

Reviewer: Exactly that's my thinking as well concerning 8.8 and the current script checks whether C8 exists and deletes the Jars, replaces with the new Jars and overwrites the configurations.

So configuration ultimately is up to the customer to be adjusted on upgrades as we only provide the base layer to make the setup work but e.g. a customer can add more additions etc.

I'm fine with either describing the assumed flow that we both have and is implemented (e.g. to also cover patch releases) or as you mention postpone till 8.8.
My feeling would be we can already include it, so people know how to deal with patch releases and revisit with 8.8.

But yes, big disclaimer that you can't upgrade from 8.6 etc.

---
id: aws-ec2
title: "Amazon EC2"
description: "Learn how to install Camunda 8 on AWS EC2 instances."
---

This guide provides a detailed walkthrough for installing the Camunda 8 single JAR on AWS EC2 instances. It focuses on managed services by AWS and their cloud offering. Finally, you will verify that the connection to your Self-Managed Camunda 8 environment is working.

This guide focuses on setting up the [orchestration cluster](/self-managed/reference-architecture/reference-architecture.md#orchestration-cluster-vs-web-modeler-and-console) for Camunda 8. The Web Modeler and Console are not covered in this manual deployment approach. These components are supported on Kubernetes and should be [deployed using Kubernetes](/self-managed/setup/install.md/#install-web-modeler).

> ## Disclaimer

Author: I think we can eliminate this header, since it's just for the note, and doesn't necessarily need the legal weight of a disclaimer? 

If the header feels necessary, I would remove the content from a note, and retitle the section to "Use other cloud providers".

:::note Using other Cloud provider

> This guide is not limited to AWS but is built around the available tools and services that AWS offers. The scripts and ideas behind it can easily be adjusted for any other cloud provider and use case.
> 
> However, if you wish to use this guide with a different cloud provider, please note that you will be responsible for configuring and maintaining the resulting infrastructure. Our support is limited to questions related to the guide itself, not to the specific tools and services of the chosen cloud provider.

Author: ```suggestion
This guide is built around the available tools and services that AWS offers, but is not limited to AWS. The scripts and ideas included can be adjusted for any other cloud provider and use case.

When using this guide with a different cloud provider, note that you will be responsible for configuring and maintaining the resulting infrastructure. Our support is limited to questions related to the guide itself, not to the specific tools and services of the chosen cloud provider.
```

    - a NAT Gateway that allows the private EC2 instances to reach the internet to download and update software packages. This can not be used to access the EC2 instances.
- [Security Groups](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-security-groups.html) to handle traffic flow to the VMs.
- an [Internet Gateway](https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Internet_Gateway.html) as glue component between the VPC and the Internet, to allow traffic between those.

Both types of subnets are distributed over three availability zones of a single AWS region, allowing for a highly available setup.

:::note Single Deployment
Alternatively, one can run the same setup with a single AWS EC2 instance as well but should be aware that in case of a zone failure, their whole setup would be unreachable.
:::

> ## Prerequisites

Author: Could this be "Requirements" to match the other guides?

- An AWS account to create any resources within AWS.
  - On a high level, permissions are required on the **ec2**, **iam**, **elasticloadbalancing**, **kms**, **logs**, and **es** level.
  - For a more fine-grained view of the permissions, check this [example policy](https://github.com/camunda/camunda-deployment-references/blob/main/aws/ec2/example/policy.json).
- Terraform (1.7+)
- Unix based Operating System (OS) with ssh and sftp
  - Windows may be used with [Cygwin](https://www.cygwin.com/) or [Windows WSL](https://learn.microsoft.com/en-us/windows/wsl/install) but has not been tested

### Considerations

- TODO: reference official limitations if there are any

> - The Optimize importer is not highly available and must only run once within the whole setup.

Author: Is this complete? Should we hide the TODO until anything else gets added in?

## Outcome

The outcome is a fully working Camunda orchestration cluster running in a high availability setup using AWS EC2 and utilizing a managed OpenSearch domain.
The EC2 instances come with an extra disk meant for Camunda to ensure that the content is separated from the operating system.

> ## Get started

Author: I think we can remove "Get started," as this h2 encompasses the entire section, and bump up the following H3s to H2s instead.

1. **Download the reference architecture GitHub repository** to directly reuse and extend the existing Terraform example base.
   The idea is to provide an easy to use example implementation that's flexible to extend to your own needs without the potential limitations of a Terraform module.

```sh
wget https://github.com/camunda/camunda-deployment-references/archive/refs/heads/main.zip

> ```

Author: This section can be moved below "infrastructure"

> ### Infrastructure

Author: More descriptive - "Create required infrastructure" ?

:::warning Cost management
Following this guide will incur costs on your Cloud provider account, namely for the EC2 instances, and OpenSearch. More information can be found on AWS and their [pricing calculator](https://calculator.aws/#/) as the total cost varies per region.

> To get an estimate, you can refer to this [example calculation](https://calculator.aws/#/estimate?id=8ce855e2d02d182c4910ec8b4ea2dbf42ea5fd1d), which can be further optimized to suit your specific use cases.

Reviewer: @conceptualshark added an example cost calculator for the setup, let me know when it makes sense to add this.
